allensdk/brain_observatory/behavior/behavior_session.py

```diff
@@ -27,108 +38,191 @@
             stimulus_timestamps=stimulus_timestamps,
             running_acquisition=running_acquisition,
             raw_running_speed=raw_running_speed,
             running_speed=running_speed,
-            metadata=metadata
+            metadata=metadata,
+            licks=licks,
+            rewards=rewards,
+            stimuli=stimuli,
+            task_parameters=task_parameters,
+            trials=trials,
+            date_of_acquisition=date_of_acquisition
         )
 
     @classmethod
-    def from_internal(cls, behavior_session_id: int) -> "BehaviorSession":
-        lims_db = db_connection_creator(
-            fallback_credentials=LIMS_DB_CREDENTIAL_MAP
-        )
+    def from_lims(cls, behavior_session_id: int,
+                  lims_db: Optional[PostgresQueryMixin] = None,
+                  stimulus_timestamps: Optional[StimulusTimestamps] = None,
+                  monitor_delay: Optional[float] = None,
+                  date_of_acquisition: Optional[DateOfAcquisition] = None) \
+            -> "BehaviorSession":
+        """
+
+        Parameters
+        ----------
+        behavior_session_id
+            Behavior session id
+        lims_db
+            Database connection. If not provided will create a new one.
+        stimulus_timestamps
+            Stimulus timestamps. If not provided, will calculate stimulus
+            timestamps from stimulus file.
+        monitor_delay
+            Monitor delay. If not provided, will use an estimate.
+        date_of_acquisition
+            Date of acquisition. If not provided, will read from
+            behavior_sessions table.
+        Returns
+        -------
+        `BehaviorSession` instance
+        """
+        if lims_db is None:
+            lims_db = db_connection_creator(
+                fallback_credentials=LIMS_DB_CREDENTIAL_MAP
+            )
 
         behavior_session_id = BehaviorSessionId(behavior_session_id)
         stimulus_file = StimulusFile.from_lims(
             db=lims_db, behavior_session_id=behavior_session_id.value)
-        stimulus_timestamps = StimulusTimestamps.from_stimulus_file(
-            stimulus_file=stimulus_file
-        )
+        if stimulus_timestamps is None:
+            stimulus_timestamps = StimulusTimestamps.from_stimulus_file(
+                stimulus_file=stimulus_file)
         running_acquisition = RunningAcquisition.from_lims(
             lims_db, behavior_session_id.value
         )
         raw_running_speed = RunningSpeed.from_lims(
-            lims_db, behavior_session_id.value, filtered=False
+            lims_db, behavior_session_id.value, filtered=False,
+            stimulus_timestamps=stimulus_timestamps
         )
         running_speed = RunningSpeed.from_lims(
-            lims_db, behavior_session_id.value
+            lims_db, behavior_session_id.value,
+            stimulus_timestamps=stimulus_timestamps
         )
         behavior_metadata = BehaviorMetadata.from_internal(
```

Comment:
The intent of `from_internal` is that we read from multiple data sources, ie LIMS DB, eye tracking file, stimulus file, etc. Calling it `from_lims` seems misleading to me since we are only reading the _path_ to the data from the LIMS DB, although maybe `from_internal` is not clear.

Reply:
This has the feeling of a conversation to which I am coming late. I am sorry for that.

My impression (which I only recently gained) is that "LIMS" refers to our entire internal data management system, of which the PostGres database is a part. If I am correct in that, data files on isilon are still a part of LIMS. We use the database to associate their paths with experiment IDs, etc.

I can imagine a time in the not-too-distant future in which it will be difficult to remember which classes get instantiated with `from_lims` and which get instantiated `from_internal`.

Looking at all of the implementations from `from_internal`,
```
cell_specimens/cell_specimens.py:    def from_internal(cls, ophys_experiment_id: int,
metadata/behavior_metadata/behavior_metadata.py:    def from_internal(
metadata/ophys_experiment_metadata/ophys_experiment_metadata.py:    def from_internal(
metadata/ophys_experiment_metadata/imaging_plane.py:    def from_internal(cls, ophys_experiment_id: int,
metadata/ophys_experiment_metadata/multi_plane_metadata/multi_plane_metadata.py:    def from_internal(
metadata/behavior_ophys_metadata.py:    def from_internal(cls, ophys_experiment_id: int,
projections.py:    def from_internal(cls, ophys_experiment_id: int,
```
They all seem to take as an argument a connection to the PostGres database and one or two other arguments. I think this means they should be `from_lims`. True: the database is mostly used to find the path to well known datafile, but I think, in a lot of cases, tracking well known file paths is what the database is for.

We should probably consult with others to see if my "LIMS is more than just a database" impression is correct. I'll ask on Slack.

Reply:
(For the record, I agree that user-facing classes are going to need to preserve `from_lims`, regardless. I think that invocation has become ingrained with our internal user community, and I don't think we want to deal with the complications that will result if we change that out from under them)